### PR TITLE
ListFeatureCollection bounding box fixes [GEOT-5682][GEOT-5684], MemoryDataStore fix [GEOT-5683]

### PR DESCRIPTION
**Pull request by goudine**

### CODE REVIEW

**Comment by jodygarnett** [comment]
Would normally pull GEOT-5683 into a seperate pull request, but since they are individual commits I can still back-port easily.

File: modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java [hunk 1]
```diff
@@ -26,36 +26,38 @@
 import java.util.Map;
 import java.util.logging.Level;
 
-import org.geotools.data.DataSourceException;
-import org.geotools.data.FeatureReader;
-import org.geotools.data.Query;
+import org.geotools.data.*;
```
Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java]: Looks like your IDE is messing with imports, we try and keep such noise out of commits so we can focus on what is going on.

File: modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java [hunk 2]
```diff
@@ -26,36 +26,38 @@
...
 import org.opengis.feature.IllegalAttributeException;
 import org.opengis.feature.simple.SimpleFeature;
 import org.opengis.feature.simple.SimpleFeatureType;
 import org.opengis.feature.type.Name;
+import org.opengis.filter.identity.FeatureId;
 
 /**
  * This is an example implementation of a DataStore used for testing.
- * 
+ *
```
Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java]: white space change?

File: modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java [hunk 3]
```diff
@@ -131,7 +133,7 @@ public void addFeatures(FeatureReader <SimpleFeatureType, SimpleFeature> reader)
             throw new DataSourceException("Problem using reader", e);
         }
         finally {
-        	reader.close();
+            reader.close();
```
Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java]: thanks for fixing tabs to space :P

File: modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java [hunk 4]
```diff
@@ -297,25 +299,50 @@ private void addFeatureInternal(SimpleFeature feature) {
...
     protected ContentFeatureSource createFeatureSource(ContentEntry entry, Query query) {
-        return new MemoryFeatureStore(entry, query);
+        return new MemoryFeatureStore(entry, query) {
+            @Override protected QueryCapabilities buildQueryCapabilities() {
+                return new QueryCapabilities() {
+                    @Override public boolean isUseProvidedFIDSupported() {
+                        return true;
+                    }
+                };
+            }
```
Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java]: please make these change directly to MemoryFeatureStore, no need for an anonymous subclass as MemoryDataStore is the "owner" of MemoryFeatureStore.

File: modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java [hunk 5]
```diff
@@ -297,25 +299,50 @@ private void addFeatureInternal(SimpleFeature feature) {
...
+
+            //need to actually use the provided FID
+            @Override protected FeatureWriter<SimpleFeatureType, SimpleFeature> getWriterInternal(
+                Query query, int flags) throws IOException {
+                return new MemoryFeatureWriter(getState(), query) {
+                    @Override public void write() throws IOException {
```
Comment (jodygarnett) [inline modules/library/data/src/main/java/org/geotools/data/memory/MemoryDataStore.java]: Same here, please fix MemoryFeatureWriter directly, no need for anonymous subclass.

File: modules/library/main/src/main/java/org/geotools/data/collection/ListFeatureCollection.java [hunk 6]
```diff
@@ -185,14 +175,16 @@ public synchronized ReferencedEnvelope getBounds() {
      * @return 
      */
     protected ReferencedEnvelope calculateBounds() {
-        ReferencedEnvelope extent = new ReferencedEnvelope();
+        ReferencedEnvelope extent = ReferencedEnvelope.create( getSchema().getCoordinateReferenceSystem() );
         for( SimpleFeature feature : list ){
-            if( feature == null ) continue;
+            if( feature == null )
+                continue;
```
Comment (jodygarnett) [inline modules/library/main/src/main/java/org/geotools/data/collection/ListFeatureCollection.java]: we tend to use braces when placing a statement on a seperate line, check developers guide for sun-java-conventions + 100 cols.

File: modules/library/main/src/test/java/org/geotools/data/collection/ListFeatureCollectionTest.java [hunk 7]
```diff
@@ -131,6 +141,65 @@ public void iterator() {
         assertTrue(copy.isEmpty());
     }
 
+    @Test
+    public void removeAndAddFeatureBounds() {
```
Comment (jodygarnett) [inline modules/library/main/src/test/java/org/geotools/data/collection/ListFeatureCollectionTest.java]: please cite bug number in javadoc; it helps with traceability.

This is a suggestion only, something we try and do to help others.

File: modules/library/main/src/test/java/org/geotools/data/collection/ListFeatureCollectionTest.java [hunk 8]
```diff
@@ -131,6 +141,65 @@ public void iterator() {
...
+        BoundingBox3D envelope1 = new ReferencedEnvelope3D( 0, 50, 0, 50, 0, 70, null);
+        Filter bbox1 = ff.bbox(ff.property("geom"), envelope1);
+
+        // initialize FC
+        featureCollection = new ListFeatureCollection(TYPE, featureList);
+
+        // add created features to the collection
+        featureCollection.add(f1);
+        featureCollection.add(f2);
+        featureCollection.add(f3);
```
Comment (jodygarnett) [inline modules/library/main/src/test/java/org/geotools/data/collection/ListFeatureCollectionTest.java]: I think we want to check the featureCollection.getBounds() method to ensure it is a ReferencedEnvelope3D?

File: modules/library/main/src/test/java/org/geotools/data/collection/ListFeatureCollectionTest.java [hunk 9]
```diff
@@ -131,6 +141,65 @@ public void iterator() {
...
+
+        // ensure features are contained in BBox
+        assertTrue(bbox1.evaluate(f1));
+        assertTrue(bbox1.evaluate(f2));
+        assertTrue(bbox1.evaluate(f3));
```
Comment (jodygarnett) [inline modules/library/main/src/test/java/org/geotools/data/collection/ListFeatureCollectionTest.java]: this just checks the behaviour of ReferencedEnvelope3D and BBOX, not ListFeatureCollection.

**Comment by aaime** [comment]
Wrong branch, should target master

**Comment by mtCarto** [comment]
@aaime This doesn't need to hold up the release, as we plan on backporting to satisfy our requirements. Will target master once we've updated as per Jody's review

**Comment by goudine** [comment]
Re-issued PR to master, with the comments addressed.
Links to the submitted PR's are:
https://github.com/geotools/geotools/pull/1521
https://github.com/geotools/geotools/pull/1522

**Pull request closed without merging**